packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -0,0 +1,129 @@
+            (value: ConnectionState, oldState: ConnectionState, reason?: string | undefined) => void,
+        private readonly propagateConnectionState: () => void,
+        private readonly isContainerLoaded: () => boolean,
+        private readonly logger: ITelemetryLogger,
+    ) {
+    }
+
+    public receivedAddMemberEvent(clientId: string, quorum: Quorum) {
+        // This is the only one that requires the pending client ID
+        if (clientId === this.pendingClientId) {
+            this.setConnectionState(ConnectionState.Connected);
+        }
+    }
+
+    public receivedDisconnectEvent(reason: string) {
+        this.setConnectionState(ConnectionState.Disconnected, reason);
+    }
+
+    public receivedConnectEvent(
+        emitter: EventEmitter,
+        connectionMode: ConnectionMode,
+        details: IConnectionDetails,
+        opsBehind?: number,
+    ) {
+        const oldState = this._connectionState;
+        this._connectionState = ConnectionState.Connecting;
+
+        // Stash the clientID to detect when transitioning from connecting (socket.io channel open) to connected
+        // (have received the join message for the client ID)
+        // This is especially important in the reconnect case. It's possible there could be outstanding
+        // ops sent by this client, so we should keep the old client id until we see our own client's
+        // join message. after we see the join message for out new connection with our new client id,
+        // we know there can no longer be outstanding ops that we sent with the previous client id.
+        this._pendingClientId = details.clientId;
+
+        emitter.emit(connectEventName, opsBehind);
+
+        // Report telemetry after we set client id!
+        this.logConnectionStateChangeTelemetry(ConnectionState.Connecting, oldState);
+
+        const protocolHandler = this.protocolHandler();
+        // Check if we already processed our own join op through delta storage!
+        // we are fetching ops from storage in parallel to connecting to ordering service
+        // Given async processes, it's possible that we have already processed our own join message before
+        // connection was fully established.
+        // Note that we might be still initializing quorum - connection is established proactively on load!
+        if ((protocolHandler !== undefined && protocolHandler.quorum.has(details.clientId))
+            || connectionMode === "read"
+        ) {
+            this.setConnectionState(ConnectionState.Connected);
+        }
+    }
+
+    private setConnectionState(value: ConnectionState.Disconnected, reason: string);
+    private setConnectionState(value: ConnectionState.Connecting | ConnectionState.Connected);
+    private setConnectionState(
+        value: ConnectionState,
+        reason?: string,
+    ) {
+        assert(value !== ConnectionState.Connecting, "Trying to set connection state while container is connecting!");
```

Comment:
We should remove this assert and change signature of this function - it is never called with Connecting value

Reply:
done